crates/ide-diagnostics/src/handlers/unresolved_method.rs

```diff
@@ -289,6 +333,117 @@ fn main() {
         );
     }
 
+    #[test]
+    fn test_add_method_fix_ref_self() {
```

Comment:
this test currently fails. we're calling a non-existent method on `&Dolphin` (not `Dolphin`) so no fix is generated. In such a case we should look for an `impl Dolphin` block and add a method that takes `&self`, or look for an `impl &Dolphin` block and add a method that takes `self`. What if both exist?